webapp/tests/karma/ts/services/training-cards.service.spec.ts

```diff
@@ -0,0 +1,460 @@
+  });
+
+  it('should do nothing if user is admin', async () => {
+    sessionService.isDbAdmin.returns(true);
+
+    service.initTrainingCards();
+
+    expect(sessionService.userCtx.callCount).to.equal(0);
+    expect(xmlFormsService.subscribe.callCount).to.equal(0);
+    expect(localDb.allDocs.callCount).to.equal(0);
+    expect(globalActions.setTrainingCard.callCount).to.equal(0);
+    expect(modalService.show.callCount).to.equal(0);
+    expect(consoleErrorMock.callCount).to.equal(0);
+  });
+
+  it('should show uncompleted training form based on user role', async () => {
+    sessionService.userCtx.returns({ roles: [ 'role_a' ], name: 'a_user' });
+    localDb.allDocs.resolves({ rows: [
+      { doc: { form: 'training:form-b' } },
+    ]});
+    clock = sinon.useFakeTimers(1653312565642); // 23/05/2022 20:29:25
+    const xforms = [
+      {
+        _id: 'abc-123',
+        internalId: 'training:form-a',
+        context: {
+          start_date: 1653139765642, // 21/05/2022 20:29:25
+          duration: 3,
+          user_roles: [ 'role_b', 'role_c' ],
+        },
+      },
+      {
+        _id: 'abc-456',
+        internalId: 'training:form-b',
+        context: {
+          start_date: 1652880565642, // 18/05/2022 20:29:25
+          duration: 2,
+          user_roles: [ 'role_a', 'role_c' ],
+        },
+      },
+      {
+        _id: 'abc-789',
+        internalId: 'training:form-c',
+        context: {
+          start_date: 1653744565642, // 28/05/2022 20:29:25
+          duration: 6,
+          user_roles: [ 'role_a' ],
+        },
+      },
+      {
+        _id: 'abc-098',
+        internalId: 'training:form-d',
+        context: {
+          start_date: 1653139765642, // 21/05/2022 20:29:25
+          duration: 19,
+        },
+      },
+      {
+        _id: 'abc-098',
+        internalId: 'training:form-e',
+        context: {
+          start_date: 1653139765642, // 21/05/2022 20:29:25
+          duration: 9,
+          user_roles: [ 'role_a', 'role_c' ],
+        },
+      },
+    ];
+    service.initTrainingCards();
+
+    expect(xmlFormsService.subscribe.calledOnce).to.be.true;
+    expect(xmlFormsService.subscribe.args[0][0]).to.equal('TrainingCards');
+    expect(xmlFormsService.subscribe.args[0][1]).to.deep.equal({ trainingForms: true, contactForms: false });
+    const callback = xmlFormsService.subscribe.args[0][2];
+
+    await callback(null, xforms);
+
+    expect(sessionService.userCtx.calledOnce).to.be.true;
+    expect(localDb.allDocs.calledOnce).to.be.true;
+    expect(localDb.allDocs.args[0][0]).to.deep.equal({
+      include_docs: true,
+      startkey: 'training:a_user',
+      endkey: 'training:a_user:\ufff0',
+    });
+    expect(globalActions.setTrainingCard.calledOnce);
+    expect(globalActions.setTrainingCard.args[0]).to.have.members([ 'training:form-e' ]);
+    expect(modalService.show.calledOnce).to.be.true;
+    expect(consoleErrorMock.callCount).to.equal(0);
+  });
```

Comment:
Please add a unit test where there are no training forms. 
Please also add a unit test where all training forms are expired. Please also add a unit test where all training forms start dates are in the future.

Reply:
Good suggestion! done!